pkg/engine/plugins.go

```diff
@@ -0,0 +1,185 @@
+		if err != nil {
+			return set, err
+		}
+		logging.V(preparePluginLog).Infof(
+			"gatherPluginsFromSnapshot(): plugin %s %s is required by first-class provider %q", pkg, version, urn)
+		set.Add(workspace.PluginInfo{
+			Name:    pkg.String(),
+			Kind:    workspace.ResourcePlugin,
+			Version: version,
+		})
+	}
+	for _, plug := range target.Snapshot.Manifest.Plugins {
+		logging.V(preparePluginLog).Infof(
+			"gatherPluginsFromSnapshot(): plugin %s %s is required by snapshot manifest", plug.Name, plug.Version)
+		set.Add(plug)
+	}
+	return set, nil
+}
+
+// ensurePluginsAreInstalled inspects all plugins in the plugin set and, if any plugins are not currently installed,
+// uses the given backend client to install them. Installations are processed in parallel, though
+// ensurePluginsAreInstalled does not return until all installations are completed.
+func ensurePluginsAreInstalled(client deploy.BackendClient, plugins pluginSet) error {
+	if client == nil {
+		logging.V(preparePluginLog).Infoln("ensurePluginsAreInstalled(): skipping due to nil client")
+		return nil
+	}
+	logging.V(preparePluginLog).Infof("ensurePluginsAreInstalled(): beginning")
+	var installTasks errgroup.Group
+	for _, plug := range plugins.Values() {
+		_, path, err := workspace.GetPluginPath(plug.Kind, plug.Name, plug.Version)
+		if err == nil && path != "" {
+			logging.V(preparePluginLog).Infof(
+				"ensurePluginsAreInstalled(): plugin %s %s already installed", plug.Name, plug.Version)
+			continue
+		}
+
+		// Launch an install task asynchronously and add it to the current error group.
+		info := plug // don't close over the loop induction variable
+		installTasks.Go(func() error {
+			logging.V(preparePluginLog).Infof(
+				"ensurePluginsAreInstalled(): plugin %s %s not installed, doing install", info.Name, info.Version)
+			return installPlugin(client, info)
+		})
+	}
+
+	err := installTasks.Wait()
+	logging.V(preparePluginLog).Infof("ensurePluginsAreInstalled(): completed")
+	return err
+}
+
+// ensurePluginsAreLoaded ensures that all of the plugins in the given plugin set that match the given plugin flags are
+// loaded.
+func ensurePluginsAreLoaded(plugctx *plugin.Context, plugins pluginSet, kinds plugin.Flags) error {
+	return plugctx.Host.EnsurePlugins(plugins.Values(), kinds)
+}
+
+// installPlugin installs a plugin from the given backend client.
+func installPlugin(client deploy.BackendClient, plugin workspace.PluginInfo) error {
+	contract.Assert(client != nil)
+	logging.V(preparePluginLog).Infof("installPlugin(%s, %s): beginning install", plugin.Name, plugin.Version)
+	if plugin.Kind == workspace.LanguagePlugin {
+		logging.V(preparePluginLog).Infof(
+			"installPlugin(%s, %s): is a language plugin, skipping install", plugin.Name, plugin.Version)
+		return nil
+	}
+
+	logging.V(preparePluginVerboseLog).Infof(
+		"installPlugin(%s, %s): initiating download", plugin.Name, plugin.Version)
+	stream, err := client.DownloadPlugin(context.TODO(), plugin)
```

Comment:
NIT: Should we file an item to track actually threading in a real context here?

Reply:
done.